packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -1425,12 +1428,18 @@ export function simplify<T>(
 						for (const field of simplifiedSpreadType.fields) {
 							addField(field.key, field)
 						}
-					} else {
-						const type = McdocType.toString(simplifiedSpreadType)
-						ctx.logger.warn(
-							`Tried to spread non-struct type ${type}`,
-						)
+					} else if (simplifiedSpreadType.kind === 'union') {
+						for (const member of simplifiedSpreadType.members) {
+							if (member.kind === 'struct') {
+								for (const field of member.fields) {
+									// TODO: technically, if a field is required in all members of
+									// this union, it could be added as a required field
+									addField(field.key, { ...field, optional: true })
```

Comment:
This is a bit problematic as it really should be resolved as another union.

E.g. if you have sth like this
```rs
struct {
    ...( struct { foo: string } | struct { bar: string } ),
    baz?: boolean
}
```

The way you are doing it here, both `foo` and `bar` will be optional so an empty struct would be valid and not raise an error, while the expected behavior would be that it reports `missing one of the keys "foo" or "bar"`.

So if we want to handle to handle unions here, the only way is to resolve them like this in the above example:
```rs
(
    struct { foo: string, baz?: bool }
    | struct { bar: string, baz?: bool }
)
```

This would also automatically take care of the TODO you provided here.

Reply:
Hmm you're right that would be an issue. Do we have any such cases in vanilla-mcdoc? I worry about exploding the size of the returned "simplified" type

Reply:
The only cases where imo we want to spread a union of structs is for entity, block, and item NBT. And in all those cases all fields are optional.

Reply:
This will as implemented right now be used for any dispatcher fallback case, for example when no known id is supplied somewhere.

There are definitely dispatched values that have required fields.

Reply:
That's why I think for most dispatcher fallbacks (except for the above 3 cases), we should not return a union of all. If the type is not available we should return an empty union.

Reply:
I'm not sure on this. The main reason it is done differently is that the [mcdoc reference](https://spyglassmc.com/user/mcdoc/#indexing-on-a-type) says the default value should be fallback.

> The default value used for all cases (including the two mutable special keys, `%none` and `%unknown`) is the [fallback case](https://spyglassmc.com/user/mcdoc/#dispatcher-fallback-case).